pkg/compact/downsample/downsample.go

```diff
@@ -256,41 +282,121 @@ type aggregator struct {
 }
 
 // reset the stats to start a new aggregation window.
-func (a *aggregator) reset() {
-	a.count = 0
-	a.sum = 0
-	a.min = math.MaxFloat64
-	a.max = -math.MaxFloat64
+func (f *floatAggregator) reset() {
+	f.count = 0
+	f.sum = 0
+	f.min = math.MaxFloat64
+	f.max = -math.MaxFloat64
 }
 
-func (a *aggregator) add(v float64) {
-	if a.total > 0 {
-		if v < a.last {
+func (f *floatAggregator) add(s sample) {
+	if f.total > 0 {
+		if s.v < f.last {
 			// Counter reset, correct the value.
-			a.counter += v
-			a.resets++
+			f.counter += s.v
+			f.resets++
 		} else {
 			// Add delta with last value to the counter.
-			a.counter += v - a.last
+			f.counter += s.v - f.last
 		}
 	} else {
 		// First sample sets the counter.
-		a.counter = v
+		f.counter = s.v
 	}
-	a.last = v
+	f.last = s.v
 
-	a.sum += v
-	a.count++
-	a.total++
+	f.sum += s.v
+	f.count++
+	f.total++
 
-	if v < a.min {
-		a.min = v
+	if s.v < f.min {
+		f.min = s.v
+	}
+	if s.v > f.max {
+		f.max = s.v
 	}
-	if v > a.max {
-		a.max = v
+}
+
+func (f *floatAggregator) processedSamples() int {
+	return f.total
+}
+
+type histogramAggregator struct {
+	total         int                       // Total histograms processed.
+	count         int                       // Histograms in current window.
+	sum           *histogram.FloatHistogram // Value sum of current window (for gauge histograms).
+	counter       *histogram.FloatHistogram // Total counter state since beginning (for counter histograms).
+	resets        int                       // Number of counter resets since beginning.
+	previous      *histogram.FloatHistogram // Previously added value.
+	schema        int32                     // Smallest schema in the batch that's being aggregated.
+	resetDetected bool                      // Whether a reset was detected since the beginning.
+}
+
+func newHistogramAggregator(schema int32) *histogramAggregator {
+	return &histogramAggregator{
+		schema: schema,
 	}
 }
 
+func (h *histogramAggregator) reset() {
+	h.count = 0
+	h.sum = nil
+}
+
+func (h *histogramAggregator) add(s sample) {
+	fh := s.fh
+	if fh.Schema < h.schema {
+		panic("schema should be greater or equal to aggregator schema")
+	}
+
+	// If schema of the sample is greater than the
+	// aggregator schema, we need to reduce the resolution.
+	if fh.Schema > h.schema {
+		fh = fh.CopyToSchema(h.schema)
+	}
+
+	if h.total > 0 {
+		if fh.DetectReset(h.previous) {
```

Comment:
The increase of schema is treated like a reset:

```
A decrease of the ZeroThreshold or an increase of the Schema (i.e. an increase of resolution) can only happen together with a reset. Thus, the method returns true in either case.
```

If we do `fh.CopyToSchema(h.schema)` previously then this case is not caught AFAICT. Shouldn't we do the same reset logic in the preceding `if`?

Reply:
I have to check that, but not sure if this will be still relevant. I proposed a change to how we can do downsampling: https://github.com/thanos-io/thanos/issues/6554

Reply:
I added downsampling support for mixed chunk series and this PR is still open for review since #6554 will be a longer term task.

Reply:
> If we do `fh.CopyToSchema(h.schema)` previously then this case is not caught AFAICT. Shouldn't we do the same reset logic in the preceding `if`?

I think you are right and will look into it.

Reply:
> If we do `fh.CopyToSchema(h.schema)` previously then this case is not caught AFAICT. Shouldn't we do the same reset logic in the preceding `if`?

`h.schema` is the precalculated minimal schema from all samples of the series, which we need to know, otherwise we might run in the situation where we can't add a sample to aggregates. For the reset we need to compare current and previous histogram, so I suggest to keep the original histogram and use that for reset detection: https://github.com/thanos-io/thanos/pull/6350/commits/d7d2224ba9811901ebee5f8c4abf59f23b0749f1

From what I understand [the head appender sets reset hints properly](https://github.com/prometheus/prometheus/blob/cbb69e51423565ec40f46e74f4ff2dbb3b7fb4f0/tsdb/head_append.go#L1316-L1328), so even without the fix it should work.